python/cudf/cudf/core/udf/typing.py

```diff
@@ -339,6 +339,36 @@ def generic(self, args, kws):
             return nb_signature(types.boolean, MaskedType(types.boolean))
 
 
+@cuda_decl_registry.register_global(float)
+class MaskedScalarFloatCast(AbstractTemplate):
+    """
+    Typing for float(Masked)
+    returns the result of calling "float" on the input
+    TODO: retains the validity of the input rather than
```

Comment:
Is this TODO something we can act on, or should we instead mark it as an difference in from the pandas behavior? e.g. "Unlike pandas, this cast retains the validity of the input instead of raising an error in the case of `float(pd.NA)`.

Also does this behavior in pandas vary depending on the selected engine?

Reply:
There's a longstanding [issue](https://github.com/rapidsai/cudf/issues/8774) around doing this that it might be time we tackle. The short answer is yes, we it seems there's a way of raising in this case although it could come with the price of some performance. It's probably out of scope of this PR. 

By engine I assume you are referring to the ability to pass `engine="numba"` to several pandas APIs. TLDR I wouldn't expect this to work at all. AFAIK pandas doesn't do any extending of numba, so I wouldn't expect to be able to use pandas to JIT any python code that involves any objects that aren't pre-built into numba (basically scalars and arrays on the CPU). This includes `pd.NA`.

 For context, here is [the NAType we use to back instances of pd.NA and cudf.NA](https://github.com/rapidsai/cudf/blob/branch-22.10/python/cudf/cudf/core/udf/typing.py#L186) in our numba extension for scalar UDFs. Without this, numba would error and complain there's something unidentified in the UDF if `NA` appeared anywhere. Pandas would also have to implement the kinds of operations and statements that would commonly appear in a UDF, such as `x is NA` like we have [here.](https://github.com/rapidsai/cudf/blob/branch-22.10/python/cudf/cudf/core/udf/lowering.py#L247) 

It turns out that straight `apply` doesn't accept `engine=`, surprisingly, but `Rolling.apply` does so we can use it to see what happens in this case.

First the baseline case which works both with and without `engine="numba"`:
```python
>>> data = pd.Series(range(1000))
>>> def f(window):
...     sum = 0
...     for element in window:
...         sum += element
...     return float(sum)
... 
>>> data.rolling(10).apply(f)
0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
        ...  
995    9905.0
996    9915.0
997    9925.0
998    9935.0
999    9945.0
Length: 1000, dtype: float64
>>> data.rolling(10).apply(f, engine='numba', raw=True)
0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
        ...  
995    9905.0
996    9915.0
997    9925.0
998    9935.0
999    9945.0
Length: 1000, dtype: float64
```
Let's add some nulls to the data
```python
data[::2] = pd.NA
data[:10]
>>> data[::2] = pd.NA
>>> data[:10]
0    <NA>
1       1
2    <NA>
3       3
4    <NA>
5       5
6    <NA>
7       7
8    <NA>
9       9
dtype: object
```
In both cases, invoking the numba engine will cause a failure when the null is encountered. It is notable that this happens during the data prep phase as well, so the function never even makes it to the compilation stage. 

```python
data.rolling(10).apply(f)
TypeError: float() argument must be a string or a number, not 'NAType'
data.rolling(10).apply(f, engine='numba', raw=True)
TypeError: float() argument must be a string or a number, not 'NAType'
```
One can play around with it in several different ways but I wasn't able to obtain a sensible result.

Reply:
Fantastic answer. Thank you, I learned a lot from the linked issue and your description.

Reply:
Can confirm that pandas doesn't extend numba at all and just throws the UDF at numba to JIT.

> It turns out that straight apply doesn't accept engine=, surprisingly, but Rolling.apply does so we can use it to see what happens in this case.

Related https://github.com/pandas-dev/pandas/issues/31845. `Rolling.apply` acts essentially like `DataFrame.transform` so it was easier to implement since the shape of the result was predictable. I didn't end up implementing `engine="numba"` for `Groupby.apply` or `DataFrame.apply` because those functions can either act like a `transform` or `agg` so the unpredictability of the resulting shape was difficult to emulate. (That's why `Groupby.agg` and `Groupby.transform` only support `engine="numba"`)

Reply:
I see, yeah, that makes a lot of sense. Thanks for the context, I had wondered about that.